Sources/Build/misc.swift

```diff
@@ -32,6 +35,83 @@ extension CModule {
     }
 }
 
+extension ClangModule {
+    
+    public enum ModuleMapError: ErrorProtocol {
+        case UnsupportedIncludeLayoutForModule(String)
+    }
+    
+    public func generateModuleMap() throws {
+        
+        //Return if module map is already present
+        guard !moduleMapPath.isFile else {
+            return
+        }
+        
+        let includeDir = path
+        
+        //Generate empty module map if include dir is not present
+        guard includeDir.isDirectory else {
+            print("warning: No include directory, generating empty module map")
+            try POSIX.mkdir(includeDir)
+            try fopen(moduleMapPath, mode: .Write) { fp in
+                try fputs("\n", fp)
+            }
+            return
+        }
+        
+        let walked = walk(includeDir, recursively: false).map{$0}
+        
+        let files = walked.filter{$0.isFile && $0.hasSuffix(".h")}
+        let dirs = walked.filter{$0.isDirectory}
+        
+        if dirs.isEmpty {
+            guard !files.isEmpty else { throw ModuleMapError.UnsupportedIncludeLayoutForModule(name) }
+            try createModuleMap(.FlatHeaderLayout)
+            return
+        }
+        
+        guard let moduleHeaderDir = dirs.first where moduleHeaderDir.basename == name && files.isEmpty else {
+            throw ModuleMapError.UnsupportedIncludeLayoutForModule(name)
+        }
+        
+        let umbrellaHeader = Path.join(moduleHeaderDir, "\(name).h")
+        if umbrellaHeader.isFile {
+            try createModuleMap(.HeaderFile)
+        } else {
+            try createModuleMap(.ModuleNameDir)
+        }
+    }
+    
+    private enum UmbrellaType {
+        case FlatHeaderLayout
+        case ModuleNameDir
+        case HeaderFile
+    }
+    
+    private func createModuleMap(type: UmbrellaType) throws {
+        let moduleMap = try fopen(moduleMapPath, mode: .Write)
```

Comment:
Given that there is no performance concern here about the module map file, I suggest the this code simply create the entire module map as a String, then write it to a file. This should be cleaner than using repeated fputs() declarations.